File: scripts/database_integration_scripts/enhance_db/src/bin/08_create_writing_languages.rs
/// Create writing_languages table from extracted JSON data.
/// Creates both a reference table of languages and a mapping table.
use anyhow::Result;
use indicatif::{ProgressBar, ProgressStyle};
use rusqlite::{params, Connection};
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::io::Write;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const JSON_PATH: &str = "data/all_humans/all_human_writing_languages.json";
const TASK_LOG: &str = "task.log";

fn log(msg: &str) {
    println!("{}", msg);
    let mut f = fs::OpenOptions::new()
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", msg).unwrap();
}

fn clean_label(s: &str) -> String {
    let s = s.trim_matches('"');
    if s.ends_with("@en") {
        s[..s.len() - 3].to_string()
    } else {
        s.to_string()
    }
}

fn main() -> Result<()> {
    log("[DB] 08: Creating writing_languages table...");

    // Load writing languages JSON
    let json_str = fs::read_to_string(JSON_PATH)?;
    let human_langs: HashMap<String, Value> = serde_json::from_str(&json_str)?;
    log(&format!("[DB] Loaded writing languages for {} individuals", human_langs.len()));

    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch("PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;")?;

    // Drop existing tables
    conn.execute_batch("DROP TABLE IF EXISTS writing_languages;")?;
    conn.execute_batch("DROP TABLE IF EXISTS individual_writing_languages;")?;

    // Create reference table for languages
    conn.execute_batch(
        "CREATE TABLE writing_languages (
            id TEXT PRIMARY KEY,
            name TEXT NOT NULL,
            count INTEGER DEFAULT 0
        );"
    )?;

    // Create mapping table
    conn.execute_batch(
        "CREATE TABLE individual_writing_languages (
            wikidata_id TEXT NOT NULL,
            individual_name TEXT,
            language_id TEXT NOT NULL,
            language_name TEXT,
            PRIMARY KEY (wikidata_id, language_id)
        );"
    )?;

    // Collect all unique languages and build mapping data
    let mut lang_counts: HashMap<String, (String, i64)> = HashMap::new(); // id -> (name, count)
    let mut mappings: Vec<(String, String, String)> = Vec::new(); // (human_id, lang_id, lang_name)

    let pb = ProgressBar::new(human_langs.len() as u64);
    pb.set_style(ProgressStyle::default_bar()
        .template("{msg} [{bar:40}] {pos}/{len} ({eta})")
        .unwrap());
    pb.set_message("Processing languages");

    for (human_id, langs) in &human_langs {
        if let Some(arr) = langs.as_array() {
            for lang in arr {
                let lang_id = lang.get("id").and_then(|v| v.as_str()).unwrap_or_default().to_string();
                let lang_name = clean_label(lang.get("name").and_then(|v| v.as_str()).unwrap_or_default());

                if !lang_id.is_empty() && !lang_name.is_empty() {
                    let entry = lang_counts.entry(lang_id.clone()).or_insert((lang_name.clone(), 0));
                    entry.1 += 1;
                    mappings.push((human_id.clone(), lang_id, lang_name));
                }
            }
        }
        pb.inc(1);
    }
    pb.finish();

    log(&format!("[DB] Found {} unique languages, {} mappings", lang_counts.len(), mappings.len()));

    // Insert languages
    conn.execute_batch("BEGIN TRANSACTION;")?;
    {
        let mut stmt = conn.prepare("INSERT OR IGNORE INTO writing_languages (id, name, count) VALUES (?1, ?2, ?3)")?;
        for (id, (name, count)) in &lang_counts {
            stmt.execute(params![id, name, count])?;
        }
    }
    conn.execute_batch("COMMIT;")?;

    // Insert mappings in batches
    log("[DB] Inserting individual-language mappings...");
    let pb = ProgressBar::new(mappings.len() as u64);
    pb.set_style(ProgressStyle::default_bar()
        .template("{msg} [{bar:40}] {pos}/{len} ({per_sec}) ({eta})")
        .unwrap());
    pb.set_message("Inserting mappings");

    conn.execute_batch("BEGIN TRANSACTION;")?;
    let mut stmt = conn.prepare(
        "INSERT OR IGNORE INTO individual_writing_languages (wikidata_id, language_id, language_name) VALUES (?1, ?2, ?3)"
    )?;

    let mut batch_count = 0u64;
    for (human_id, lang_id, lang_name) in &mappings {
        stmt.execute(params![human_id, lang_id, lang_name])?;
        batch_count += 1;
        if batch_count % 100_000 == 0 {
            conn.execute_batch("COMMIT; BEGIN TRANSACTION;")?;
        }
        pb.inc(1);
    }
    conn.execute_batch("COMMIT;")?;
    pb.finish();

    // Update individual_name in the mapping table
    log("[DB] Updating individual names in writing_languages mapping...");
    conn.execute(
        "UPDATE individual_writing_languages SET individual_name = (
            SELECT individuals.name_en FROM individuals
            WHERE individuals.wikidata_id = individual_writing_languages.wikidata_id
        )",
        [],
    )?;

    // Create indexes
    conn.execute_batch("CREATE INDEX IF NOT EXISTS idx_iwl_wikidata ON individual_writing_languages(wikidata_id);")?;
    conn.execute_batch("CREATE INDEX IF NOT EXISTS idx_iwl_lang ON individual_writing_languages(language_id);")?;

    log("[DB] 08: Done. Created writing_languages tables.");
    Ok(())
}

File: scripts/database_integration_scripts/legacy/08_create_writing_languages.py
"""08 - Build writing_languages and individual_writing_languages tables.

Mirrors `enhance_db/src/bin/08_create_writing_languages.rs`.

  Inputs : data/all_humans/all_human_writing_languages.json
            { wikidata_id: [ {id, name}, ... ] }
           individuals (for individual_name backfill)
  Output : writing_languages              (id PK, name, count)
           individual_writing_languages   (wikidata_id, language_id, ...)
           Two indexes on the mapping table.

Usage
-----
    python3 08_create_writing_languages.py
    python3 08_create_writing_languages.py --full
"""

from __future__ import annotations

import json
import sqlite3
import tempfile
from pathlib import Path

from common import (
    ALL_HUMANS_DIR,
    executemany_batched,
    insert_rows,
    load_json,
    log,
    open_db,
    parse_run_mode,
)

JSON_PATH = ALL_HUMANS_DIR / "all_human_writing_languages.json"


def clean_label(s: str) -> str:
    s = s.strip().strip('"')
    if s.endswith("@en"):
        s = s[:-3]
    return s


def run(conn: sqlite3.Connection, json_path: Path = JSON_PATH) -> tuple[int, int]:
    log("[DB] 08: Creating writing_languages table...")
    human_langs = load_json(json_path)
    log(f"[DB] Loaded writing languages for {len(human_langs)} individuals")

    conn.execute("DROP TABLE IF EXISTS writing_languages")
    conn.execute("DROP TABLE IF EXISTS individual_writing_languages")
    conn.execute(
        """
        CREATE TABLE writing_languages (
            id TEXT PRIMARY KEY,
            name TEXT NOT NULL,
            count INTEGER DEFAULT 0
        )
        """
    )
    conn.execute(
        """
        CREATE TABLE individual_writing_languages (
            wikidata_id TEXT NOT NULL,
            individual_name TEXT,
            language_id TEXT NOT NULL,
            language_name TEXT,
            PRIMARY KEY (wikidata_id, language_id)
        )
        """
    )

    lang_counts: dict[str, list] = {}  # id -> [name, count]
    mappings: list[tuple[str, str, str]] = []

    try:
        from tqdm import tqdm
        items = tqdm(human_langs.items(), total=len(human_langs), desc="Processing languages")
    except ImportError:
        items = human_langs.items()

    for human_id, langs in items:
        if not isinstance(langs, list):
            continue
        for lang in langs:
            if not isinstance(lang, dict):
                continue
            lang_id = (lang.get("id") or "").strip()
            lang_name = clean_label(lang.get("name") or "")
            if not lang_id or not lang_name:
                continue
            entry = lang_counts.setdefault(lang_id, [lang_name, 0])
            entry[1] += 1
            mappings.append((human_id, lang_id, lang_name))

    log(f"[DB] Found {len(lang_counts)} unique languages, {len(mappings)} mappings")

    conn.executemany(
        "INSERT OR IGNORE INTO writing_languages (id, name, count) VALUES (?, ?, ?)",
        [(lid, name, count) for lid, (name, count) in lang_counts.items()],
    )
    conn.commit()

    log("[DB] Inserting individual-language mappings...")
    executemany_batched(
        conn,
        "INSERT OR IGNORE INTO individual_writing_languages "
        "(wikidata_id, language_id, language_name) VALUES (?, ?, ?)",
        mappings,
        batch_size=100_000,
        desc="Inserting mappings",
        total=len(mappings),
    )

    log("[DB] Updating individual names in writing_languages mapping...")
    conn.execute(
        """
        UPDATE individual_writing_languages SET individual_name = (
            SELECT individuals.name_en FROM individuals
            WHERE individuals.wikidata_id = individual_writing_languages.wikidata_id
        )
        """
    )
    conn.commit()

    conn.execute("CREATE INDEX IF NOT EXISTS idx_iwl_wikidata ON individual_writing_languages(wikidata_id)")
    conn.execute("CREATE INDEX IF NOT EXISTS idx_iwl_lang ON individual_writing_languages(language_id)")
    conn.commit()
    log("[DB] 08: Done. Created writing_languages tables.")
    return len(lang_counts), len(mappings)


def _sample_main() -> None:
    fake = {
        "Q1": [{"id": "Q1860", "name": "English"}, {"id": "Q150", "name": "French"}],
        "Q2": [{"id": "Q1860", "name": "English"}],
        "Q3": [{"id": "Q150", "name": "French@en"}],
    }
    with tempfile.TemporaryDirectory() as tmp:
        json_path = Path(tmp) / "wl.json"
        json_path.write_text(json.dumps(fake))
        db_path = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db_path) as seed:
            seed.execute("CREATE TABLE individuals (wikidata_id TEXT PRIMARY KEY, name_en TEXT)")
            insert_rows(seed, "individuals", [
                {"wikidata_id": "Q1", "name_en": "Alice"},
                {"wikidata_id": "Q2", "name_en": "Bob"},
                {"wikidata_id": "Q3", "name_en": "Cleo"},
            ])
            seed.commit()

        with open_db(db_path) as conn:
            n_langs, n_maps = run(conn, json_path=json_path)
            wl = conn.execute(
                "SELECT id, name, count FROM writing_languages ORDER BY count DESC"
            ).fetchall()
            iwl = conn.execute(
                "SELECT wikidata_id, individual_name, language_id, language_name "
                "FROM individual_writing_languages ORDER BY wikidata_id, language_id"
            ).fetchall()
        log(f"[sample] {n_langs} languages, {n_maps} mappings")
        for r in wl:
            log(f"  writing_languages: {r}")
        for r in iwl:
            log(f"  individual_writing_languages: {r}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn)
    else:
        _sample_main()
